AY2122S1-CS2103T-T11-2/tp
# Fix filter and handle corrupted json alerts

### Description

Filter conditions are now union with each other. Indices taken in for i/ and r/ prefixes can now be parse as long as there is whitespace: i.e `r/1               10` will now be parsed as range 1 to 10 

Alerts on corrupted json files will also be send to users.
3 types of errors:
1. Missing json file -> New json file with sample json will be created
2. Corrupted json file (cannot be parsed) -> Empty json file will be created
3. IOExceptions -> Empty json file will be created
4. Directory in data directory named `flashcardapp.json` -> Warning will be created telling users to delete the directory themselves

Note, this commit does not throw error if users give extraneous fields (like in issue #171 ). Since AB3 deliberately allows for extra fields that will be ignored in its implementation and tests

### Type of change



### How to test

Testing filter:
1. filter l/chinese r/1 10  should return all chinese language cards and cards within the range of 1 to 10
2. filter r/1      10 should work correctly as well

Testing corrupted jsons:
1. Delete json file in data folder and restart app, a warning should be created, and a new file with sample data is loaded
2. Restart app again, no more warning should be thrown.
3. Introduce bugs to the json file, e.g. delete brackets, commas, add null value and then restart app
4. A warning should be thrown, and empty data set is loaded.
5. Restart app again, no more warning should be thrown.

Tested json files the commit defends against:
1. Empty
2. Single `null` value in list
3. Empty braces `{}`
4. Braces with null inside:: `{null}`
5. `{ flashcard: null }`
6. `{
  "flashcards" : null
}`
7. `{
  "flashcards" : [null]
}`
8. `{
  "notFlashcard" : []
}`
9. `{
  "flashcards" : [ {
    "languageType" : "test",
    "englishPhrase" : "test",
    "foreignPhrase" : null
  } ]
}` (null foreignPhrase)
10. `{
  "flashcards" : [[][]]
}`
11. `{
  "flashcards" : [ {
    "languageType" : "test",
    "englishPhrase" : "test"
    "foreignPhrase" : "test"
  } ]
}` (missing comma)
11. Missing Json file
### Checklist



- [x] I have written testcases to cover all new methods (only for filter)
- [ ] I have updated the User Guide (update separately)
- [ ] I have updated the Developer Guide (update separately)
- [ ] I have updated my individual project portfolio
